uast/diff/diff.go

```diff
@@ -0,0 +1,283 @@
+package diff
+
+import (
+	// TODO: integrate with https://github.com/src-d/lapjv if perf is unacceptable
+	"fmt"
+	"github.com/heetch/lapjv"
+	"gopkg.in/bblfsh/sdk.v2/uast/nodes"
+)
+
+type decisionType interface {
+	isDecisionType()
+}
+
+type basicDecisionType struct{}
+
+func (_ basicDecisionType) isDecisionType() {}
+
+type decision struct {
+	cost     int
+	decision decisionType
+}
+
+// match decision types together with their params
+type sameDecision struct{ basicDecisionType }
+type replaceDecision struct{ basicDecisionType }
+type matchDecision struct{ basicDecisionType }
+type permuteDecision struct {
+	basicDecisionType
+	permutation []int
+}
+
+//convinience method for choosing the cheapest decision
+func (self *decision) minEq(candidate *decision) {
+	if _, ok := candidate.decision.(replaceDecision); ok || self.cost > candidate.cost {
+		self.cost, self.decision = candidate.cost, candidate.decision
+	}
+}
+
+//type for cache
+type keyType struct{ k1, k2 ID }
+
+//cache for diff computation
+type cacheStorage struct {
+	decisions map[keyType]decision
+	sizes     map[ID]int
+	changes   Changelist
+}
+
+func emptyCacheStorage() cacheStorage {
+	return cacheStorage{
+		make(map[keyType]decision),
+		make(map[ID]int),
+		make([]Change, 0),
+	}
+}
+
+// caches (for perf) tree size (node count)
+func (ds *cacheStorage) nodeSize(n nodes.Node) int {
+	label := nodes.UniqueKey(n)
+	if cnt, ok := ds.sizes[label]; ok {
+		return cnt
+	}
+	ret := nodes.Count(n, nodes.KindsNotNil)
+	ds.sizes[label] = ret
+	return ret
+}
+
+// find the cheapest way to naively match src and dst and return the action with its combined cost
+func (ds *cacheStorage) decideAction(src, dst nodes.Node) decision {
+	label := keyType{nodes.UniqueKey(src), nodes.UniqueKey(dst)}
+
+	if val, ok := ds.decisions[label]; ok {
+		return val
+	}
+
+	// one can always just create the dst ignoring the src
+	cost := ds.nodeSize(dst) + 1
+	bestDecision := decision{cost, replaceDecision{}}
+
+	if nodes.KindOf(src) != nodes.KindOf(dst) {
+		ds.decisions[label] = bestDecision
+		return bestDecision
+	}
+
+	switch src.(type) {
+
+	case nodes.Value:
+		src, dst := src.(nodes.Value), dst.(nodes.Value)
+		if src == dst {
+			bestDecision.minEq(&decision{0, sameDecision{}})
+		} else {
+			bestDecision.minEq(&decision{1, replaceDecision{}})
+		}
+
+	case nodes.Object:
+		src, dst := src.(nodes.Object), dst.(nodes.Object)
+		cost = 0
+
+		keys := make(map[string]bool)
+		iterate := func(keyset nodes.Object) {
+			for key := range keyset {
+				if in := keys[key]; !in {
+					keys[key] = true
+					cost += ds.decideAction(src[key], dst[key]).cost
+				}
+			}
+		}
+		iterate(src)
+		iterate(dst)
+
+		if cost == 0 {
+			bestDecision.minEq(&decision{cost, sameDecision{}})
+		} else {
+			bestDecision.minEq(&decision{cost, matchDecision{}})
+		}
+
+	case nodes.Array:
+		src, dst := src.(nodes.Array), dst.(nodes.Array)
+		cost = 0
+		if len(src) == len(dst) && func() int {
+			sum := 0
```

Comment:
Make it a proper method of `ds`, or move the code out from an anonymous function.

Reply:
it's used only once here and won't be used anywherte else. I've used the lambda only in order to leverage the lazy ifs (&& won't execute 2nd argument if 1st is false, right?) and avoid nested if construction. How to proceed then?

Reply:
I would say it should be a separate method on `ds` then. Inlining functions like this is not an idiomatic Go.